api: Create logic for automatically cleaning up dead streams

victorges opened the pull request:
**What does this pull request do? Explain your changes. (required)**

This creates an API and a corresponding cronjob to clean-up stream that are marked active
but haven't been updated in a while. This is an improvement from the lazy approach that
only tried to clean-up the streams once they were accessed via the API, which also had the
undesirable effect of triggering expensive mutations from read-only requests.

**Specific updates (required)**

- Create the new jobs API
- Create the cronjbo action
- Create a separate DB connection pool for background operations like this (and webhooks, and tasks)
- Remove old lazy logic ✨ 

**How did you test each of these updates (required)**
- [x] Ran the queries on the DB
- [x] `yarn test`
- [ ] Try it on staging

**Does this pull request close any open issues?**
Implements ENG-869

**Checklist**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

victorges committed:
api: Create API for clean-up actually inactive streams

victorges committed:
api: Move the filtering logic out of activeCleanup

victorges committed:
.github: Create cronjob to clean-up inactive streams

victorges committed:
api: Fix the active cleanup query

- We were missing the parentId check so it was reading way
more streams than necessary (all child streams/sessions).
- There are A LOT of forgotten streams in our DB (400k parents),
so we need to put some limit on this query

victorges committed:
api: Avoid using all DB resources for cleanup

victorges committed:
api: Fix tests (they were broken before!)

leszko commented on .github/workflows/cron-streams-active-cleanup.yaml:
```diff
@@ -0,0 +1,15 @@
+name: "cron: Active streams cleanup"
```
I wonder if we shouldn't run it as a kubernetes Cron Jobs. I know that currently we don't have a clear agreement on whether cron jobs should be GH Actions or Kubernetes CronJobs, but the advantages I see for integrating it in our Infra:
- Alerts integration (Grafana Pager Duty)
- (super minor) GH Actions load on the GH workers
- (minor) security and passing admin LP token

Anyway, I leave it up to you.

victorges replied:
Even though I agree it would be much better to run cronjobs, I'd leave that as a separate change for the future, given we have other (more critical) workflows already running on GitHub actions, mainly usage billing. We should probably bring this up as a tech debt to be prioritized eventually though.

leszko commented on packages/api/src/controllers/stream.ts:
```diff
@@ -229,6 +234,7 @@ function activeCleanupOne(
   }
 
   setImmediate(async () => {
+    await cleanupSemaphore.wait();
```
The "right" solution would be to have a separate connection pool for the cleanup work. This is a workaround. I wonder how difficult would it be to implement a separate connection pool for the cleanup.

victorges replied:
I've implemented the separate connection pool here!
https://github.com/livepeer/studio/pull/2013/commits/ef6df9300d01fc866ef380e41360065e29673f5d

I'd recommend reviewing that commit separately, since I touched several different files (also moved webhooks, tasks and usage logic to the jobs conn pool).

leszko commented on .github/workflows/cron-streams-active-cleanup.yaml:
```diff
@@ -0,0 +1,15 @@
+      - name: Clean-up active streams
+        run: |
+          curl --request POST \
+          --url 'https://livepeer.studio/api/stream/job/active-cleanup' \
```
Will this GH Action fail, if the cleanup fails? I think we would like to have it working this way:
1. The call `https://livepeer.studio/api/stream/job/active-cleanup` would be sync, so the GH Action effectively waits for the streams to get cleaned up
2. If it fails, the whole job should fail, so that we know that something is wrong

victorges replied:
Good call. I've ran [some tests with the GH action](https://github.com/livepeer/studio/actions/workflows/cron-streams-active-cleanup.yaml) and we can get the action to fail simply by adding. `--fail` to the curl command, so it exits with a bad code if the server responds a bad status as well. 

The clean-up implementation itself is already synchronous as well, so that part should be working once the API actually gets deployed.

leszko commented on .github/workflows/cron-streams-active-cleanup.yaml:
```diff
@@ -0,0 +1,15 @@
+
+on:
+  schedule:
+    - cron: "*/10 * * * *" # run every 10 minutes
```
How about running it every 5 min and getting rid of all the opportunistic cleanup? I think this job should be the only place where we do the stream cleanup.

Otherwise, we risk not exposing a lot of bugs + it's super wrong that making a GET call to studio makes a change in the Studio DB.

victorges replied:
Super scary, but I also would love to remove the opportunistic logic. Implemented here! https://github.com/livepeer/studio/pull/2013/commits/f70dfe3b4ff708d21eb97a596db35e45ba03013a

Notice that with a 5 minute frequency, the worst case delay for a stream to be activecleanedup will be 10 minutes. I think making the github action more frequent would not be that good though given limited runners etc, so we can consider that once we move to an actual cron in our infra. WDYT?

leszko commented on packages/api/src/util.ts:
```diff
@@ -28,21 +28,48 @@ export const sleep = (duration) => {
+  return sem;
 };
 
+export class Semaphore {
```
Did you consider using some library for this instead of implementing the Semaphore?

If we end up having our own Semaphore implementation, then how about writing some unit tests?

victorges replied:
Not necessary anymore given I'm using a separate conn pool instead :)

But FTR, I remember I _did_ search for libraries back then and surprisingly there weren't any suitable! I guess people don't really care about concurrency in node since it's single threaded 🤷‍♂️

leszko commented on packages/api/src/webhooks/cannon.test.ts:
```diff
@@ -286,7 +286,8 @@ describe("webhook cannon", () => {
       expect(receivedEvent).toBe("stream.started");
 
-      sem = semaphore();
+      // at this point the semaphore is re-acquired cause we just waited (acquired) above
```
Could you also add unit test for `/job/active-cleanup`

leszko reviewed (commented):
Good work @victorges !

Added some comments, but in general I really like the approach and love that we'll start having this functionality!

victorges committed:
Revert "api: Avoid using all DB resources for cleanup"

This reverts commit aff5120983eae098e6c5684f739ca3ae5b381794.

victorges committed:
api: Create separate DB conn pool for bg jobs

Use it on:
- Webhooks queue handler
- Tasks queue handler
- Usage billing job
- (new) Active cleanup job

victorges committed:
api: Fix tests

victorges commented on packages/api/src/controllers/webhook.test.ts:
```diff
@@ -288,8 +290,22 @@ describe("controllers/webhook", () => {
-      // const setActiveResJson = await setActiveRes.json()
-      // expect(setActiveResJson).toBeDefined()
+
+      let hooksCalled = 0;
+      for (let i = 0; i < 5 && hooksCalled < 2; i++) {
+        await sleep(500);
+
+        hooksCalled = 0;
+        for (const id of [webhookResJson.id, webhookResJson2.id]) {
+          const res = await client.get(`/webhook/${id}`);
+          expect(res.status).toBe(200);
+          const { status } = (await res.json()) as Webhook;
+          if (status.lastTriggeredAt >= now) {
+            hooksCalled++;
+          }
+        }
+      }
+      expect(hooksCalled).toBe(2);
```
FTR: this was the very cryptic failure I was getting from the tests. What has happening was that, after creating the separate jobs DB pool, the test below (`records webhook logs`) started failing because the webhook created _there_ was begin called twice, one additional time with the event triggered in _this_ test.

The problem was that this test never really waited (or even checked) that the webhooks created here were actually triggered. So the test finished while leaving the background runner calling the webhooks for it. This probably caused some flakiness every now and then as well.

What has happening was that, after switching to the separate DB pool, the very first event handled by the `WebhookCannon` took a little longer since it had to setup a new connection with Postgres (that hadn't been already initialized from the usual test setup). This small delay was already enough for the next test to start and create another webhook, which then got called for the same event. I could reproduce this by making a dummy query on the `jobsDb` in this test, which then caused the failure to stop cause the pool was then already initialized.

Fixed this by correcting this test logic tho, which was not even checking that the webhooks were being called. Explaining here only cause it seems so unrelated to the changes here, but they actually were.

victorges committed:
api: Stop triggering isActive clean-up from GETs

Still monkey-patch responses to keep the same UX,
but the actual recordings processing will only be
triggered once the active clean-up job is triggered.

victorges committed:
api: Add tests for cleanup job

victorges committed:
api: Change default limit to 1000

It's a bit too expensive to do 10000 at once

victorges committed:
api: Fix override of isActive (breaking tests)

victorges committed:
api/stream: Clean-up lost sessions as well as parents

Turns out there are also children lost from their parents out
there, so we have to list them all.

victorges committed:
api/cannon: Make sure we clear unused sessions/streams objects

victorges committed:
api/cannon: Fix typing of handleRecordingWaitingChecks

victorges committed:
api: Make sure we always clear all isActive fields

victorges committed:
api/stream: Fix clean-up streams filtering logic

victorges committed:
api: Clear streams that already have assets generated

victorges committed:
api/stream: Remove ordering from list query

Just makes things super slow and is not useful

victorges committed:
api/stream: Clean all child streams from same sess at once

victorges committed:
api/stream: Make size of clean-up more predictable

Don't filter on the outer level, but rather make sure
the pipeline supports an array of child streams.

victorges committed:
api/stream: Remove test for deduping by sessionId

Now we clean all of them, but we merge before the cleanup
processing only the session once. No easy way to catch that on
tests.

victorges committed:
api/stream: Stop using lodash

victorges committed:
api/stream: Run clean-up synchronously on API

Also parallelize parent streams

victorges committed:
api/stream: Default streamId to empty str

victorges committed:
api/stream: Parallelize child stream processing

victorges committed:
api/stream: Only use isActive field from streams

There are some bugged sessions that got isActive set to them

leszko commented on packages/api/src/controllers/stream.ts:
```diff
@@ -293,65 +294,100 @@ export function getFLVPlaybackUrl(ingest: string, stream: DBStream) {
+/**
+ * Creates an updated stream object with a fixed `isActive` field in case it is lost and needs clean up.
+ */
+function withFixedIsActive(stream: DBStream) {
```
Does it mean that we can have a stream object without `isActive` field? 

I wonder if we should fix such streams in the clean-up logic. I think the sematic of not having `isActive` field would be that the stream is not active. So, probably no need to clean it up.

victorges replied:
So this was just to keep the same logic of returning `isActive: false` after the active timeout for streams that got lost in action. Similar to what had before, but instead of triggering the clean-up job at that point we just monkey patch the API object before returning it, and wait for the clean-up job to pick it up.

I made it like this to make sure we keep retrocompatibility, tho thinking further about it I thought that we could also just not have the monkey patch at all? So we keep returning `isActive: true` for the user for up to ~7 minutes (worst case) until a clean-up job comes and fixes it (and triggers recording).

WDYT? Should we keep the retrocompatibility in the sense "if the stream is not seen in 90s we already start returning `isActive: false`", or should we just remove this monkey patch once and for all?

leszko commented on .github/workflows/cron-streams-active-cleanup.yaml:
```diff
@@ -12,6 +12,9 @@ jobs:
           - name: "staging"
             hostname: "livepeer.monster"
             api-token-secret: LP_STAGING_API_ADMIN_TOKEN
+          - name: production
```
Don't we want to run stream clean-up as Kubernetes CronJobs rather than GH Actions? I think it should be simpler to monitor with our existing infra.

victorges replied:
Yeah agreed! I don't think it's in scope of this PR though as explained here: https://github.com/livepeer/studio/pull/2013#discussion_r1579080037

leszko reviewed (approved):
Added 2 comments. Other than that, LGMT 👍

victorges committed:
Revert "api/stream: Only use isActive field from streams"

This reverts commit 862108971930382a67a23d3acdd7d1bfa540f740.

It was a staging only thing, likely from a development version.

victorges committed:
api/stream: Fallback to createdAt if lastSeen never set

victorges committed:
api/stream: Remove monkey-patching of stream objects 😍

victorges committed:
[DEV] Increase log level to debug in staging

victorges committed:
api/stream: Only trigger recording processing when stream goes offline

victorges committed:
api/stream: Check child streams shouldCleanupIsActive as well
